File: HX2.py
from parameters import *
from ode_solver import ode_solver
# New Form Parameters
C1 = V_he2_s
C2 = -U2_hx / (M_he2_s * c_p_ss)
C3 = V_he2_ss
C4 = U2_hx / (M_he2_ss * c_p_sss)

# Discretize the spatial domain
A_HX2 = (-2*np.diag(np.ones(Nx)) + np.diag(np.ones(Nx - 1), 1) + np.diag(np.ones(Nx - 1), -1)) / dx
A_HX2[0, 0] = 1 / dx
A_HX2[-1, -1] = 1 / dx

def HX2(y_hx2, Ts_HX2_L, Tss_HX2_0, step):
    
    y_hx2[:Nx,-1]=Ts_HX2_L
    y_hx2[Nx:,0]=Tss_HX2_0
    
    # y is a vector with u and v concatenated
    def pde_to_ode_hx2(t,y):
        u=y[:Nx]
        v=y[Nx:]
        
        du_dt = C1 * (A_HX2 @ u) + C2 * (u - v)
        dv_dt = C3 * (A_HX2 @ v) + C4 * (u - v)
        
        # Apply time-varying boundary conditions
        du_dt[0] = u2_L - u[0]
        du_dt[-1] = u2_H - u[-1]
        dv_dt[0] = v2_L - v[0]
        dv_dt[-1] = v2_H - v[-1]
        
        dydt = np.concatenate([du_dt, dv_dt])
        return dydt
    
    # Initial condition vector
    if step==0:
        y0 = np.concatenate([u2_init, v2_init])
    else:
        y0 = y_hx2[:,-1]
        
    # solution_y_hx1 = solve_ivp(pde_to_ode_hx1, (step, step+1), y0, t_eval=t, method='RK45')     
    bc=[]
    solution_y_hx2 = ode_solver(y0, bc, pde_to_ode_hx2)
        
    # y_hx1 is a vector at time step+1
    y_hx2 = solution_y_hx2.y   
    # print("testHX1")
    # print(y_hx1.shape)
    return y_hx2

File: src/HX2.cpp
#include "HX2.hpp"
#include "parameters.hpp"
#include "ode_solver_HX2.hpp"

void pde_to_ode_hx2(float t, const float y[length_hx], float dydt[length_hx], Param_HX2 &params) {
    // Split the input state vector y into u and v
    const float* u = &y[0];    // u is the first half of y
    const float* v = &y[Nx];   // v is the second half of y

    float du_dt[Nx] = {0};    // Array to store the derivative of u
    float dv_dt[Nx] = {0};    // Array to store the derivative of v

    // Compute du_dt and dv_dt using the provided constants
    for (int i = 0; i < Nx; ++i) {
        for (int idx = params.A_HX2_csr.row_pointers[i]; idx < params.A_HX2_csr.row_pointers[i + 1]; ++idx) {
            int j = params.A_HX2_csr.col_indices[idx];
            du_dt[i] += params.A_HX2_csr.values[idx] * u[j];
            dv_dt[i] += params.A_HX2_csr.values[idx] * v[j];
        }
        du_dt[i] += params.C2_2 * (u[i] - v[i]);
        dv_dt[i] += params.C4_2 * (u[i] - v[i]);
    }

    // Apply time-varying boundary conditions
    du_dt[0] = params.u2_L - u[0];
    du_dt[Nx - 1] = params.u2_H - u[Nx - 1];
    dv_dt[0] = 0; // Fixed condition: no change at v[0]
    dv_dt[Nx - 1] = 0; // Fixed condition: no change at v[Nx-1]

    // Populate dydt with the derivatives
    for (int i = 0; i < Nx; ++i) {
        dydt[i] = du_dt[i];        // First half for du_dt
        dydt[Nx + i] = dv_dt[i];   // Second half for dv_dt
    }
}

void HX2(float y_hx2[length_hx], float Ts_HX2_L, int step, Param_HX2 &params) {
    // Set boundary conditions
    y_hx2[Nx + 1] = params.v2_L;    // Fixed boundary condition at v[0]
    y_hx2[length_hx - 1] = params.v2_H; // Fixed boundary condition at v[L]

    float u[Nx], v[Nx];
    if (step == 0) {
        for (int i = 0; i < Nx; ++i) {
            u[i] = params.u2_init[i];
            v[i] = params.v2_init[i];
        }
    } else {
        for (int i = 0; i < Nx; ++i) {
            u[i] = y_hx2[i];
            v[i] = y_hx2[Nx + i];
        }
    }
    u[Nx - 1] = Ts_HX2_L;

    // Initial condition vector y0
    float y0[length_hx];
    if (step == 0) {
        for (int i = 0; i < Nx; ++i) {
            y0[i] = params.u2_init[i];
            y0[Nx + i] = params.v2_init[i];
        }
    } else {
        for (int i = 0; i < length_hx; ++i) {
            y0[i] = y_hx2[i];
        }
    }

    // Solve the system of ODEs
    ode_solver_hx2(y0, pde_to_ode_hx2, step, params);

    // Update y_hx2 with the solution from the ODE solver
    for (int i = 0; i < length_hx; ++i) {
        y_hx2[i] = y0[i];  // Update the original array with the new values
    }
}
